File: Tugas3/app/src/main/java/id/ac/ub/papb/recycler1/MainActivity.java
package id.ac.ub.papb.recycler1;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;

import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    RecyclerView rv1;
    public static String TAG = "RV1";
    ArrayList<Mahasiswa> data;
    MahasiswaAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rv1 = findViewById(R.id.rv1);
        EditText etNim = findViewById(R.id.etNim);
        EditText etNama = findViewById(R.id.etNama); // sebelumnya: editTextTextPersonName2
        Button bt1 = findViewById(R.id.bt1);

        data = getData();
        adapter = new MahasiswaAdapter(this, data);
        rv1.setAdapter(adapter);
        rv1.setLayoutManager(new LinearLayoutManager(this));

        bt1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String nimBaru = etNim.getText().toString().trim();
                String namaBaru = etNama.getText().toString().trim();

                if (!nimBaru.isEmpty() && !namaBaru.isEmpty()) {
                    Mahasiswa mhsBaru = new Mahasiswa();
                    mhsBaru.nim = nimBaru;
                    mhsBaru.nama = namaBaru;

                    data.add(mhsBaru);
                    adapter.notifyItemInserted(data.size() - 1);

                    // Kosongkan input
                    etNim.setText("");
                    etNama.setText("");
                }
            }
        });
    }

    public ArrayList<Mahasiswa> getData() {
        ArrayList<Mahasiswa> data = new ArrayList<>();
        List<String> nim = Arrays.asList(getResources().getStringArray(R.array.nim));
        List<String> nama = Arrays.asList(getResources().getStringArray(R.array.nama));
        for (int i = 0; i < nim.size(); i++) {
            Mahasiswa mhs = new Mahasiswa();
            mhs.nim = nim.get(i);
            mhs.nama = nama.get(i);
            Log.d(TAG,"getData "+mhs.nim);
            data.add(mhs);
        }
        return data;
    }
}

File: Tugas4/app/src/main/java/com/example/tugas4/MainActivity.kt
package com.example.tugas4

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    private lateinit var rvPosts: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        rvPosts = findViewById(R.id.rvPosts)
        rvPosts.layoutManager = LinearLayoutManager(this)

        RetrofitClient.instance.getPosts().enqueue(object : Callback<List<Post>> {
            override fun onResponse(call: Call<List<Post>>, response: Response<List<Post>>) {
                if (response.isSuccessful) {
                    val adapter = PostAdapter(response.body()!!) {
                        val intent = Intent(this@MainActivity, DetailActivity::class.java)
                        intent.putExtra("POST_ID", it.id)
                        startActivity(intent)
                    }
                    rvPosts.adapter = adapter
                }
            }

            override fun onFailure(call: Call<List<Post>>, t: Throwable) {
                Log.e("MainActivity", "Failed to fetch posts", t)
            }
        })
    }
}